main.go

```diff
@@ -0,0 +1,275 @@
+package caddy
+
+import (
+	"bytes"
+	"context"
+	"github.com/caddyserver/caddy/v2"
+	"github.com/caddyserver/caddy/v2/caddyconfig"
+	"github.com/caddyserver/caddy/v2/caddyconfig/caddyfile"
+	"github.com/caddyserver/caddy/v2/caddyconfig/httpcaddyfile"
+	"github.com/caddyserver/caddy/v2/modules/caddyhttp"
+	"github.com/darkweak/souin/cache/coalescing"
+	"github.com/darkweak/souin/configurationtypes"
+	"github.com/darkweak/souin/plugins"
+	"github.com/darkweak/souin/rfc"
+	"go.uber.org/zap"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"sync"
+)
+
+type key string
+
+const getterContextCtxKey key = "getter_context"
+const moduleName = "cache"
+const moduleID caddy.ModuleID = "http.handlers." + moduleName
+
+func init() {
+	caddy.RegisterModule(SouinCaddyPlugin{})
+	httpcaddyfile.RegisterGlobalOption(moduleName, parseCaddyfileGlobalOption)
+	httpcaddyfile.RegisterHandlerDirective(moduleName, parseCaddyfileHandlerDirective)
+}
+
+// SouinCaddyPlugin declaration.
+type SouinCaddyPlugin struct {
+	plugins.SouinBasePlugin
+	Configuration *Configuration
+	logger        *zap.Logger
+	LogLevel      string `json:"log_level,omitempty"`
+	bufPool       *sync.Pool
+	Headers       []string                           `json:"headers,omitempty"`
+	Olric         configurationtypes.CacheProvider   `json:"olric,omitempty"`
+	TTL           string                             `json:"ttl,omitempty"`
+	YKeys         map[string]configurationtypes.YKey `json:"ykeys,omitempty"`
+}
+
+// CaddyModule returns the Caddy module information.
+func (SouinCaddyPlugin) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID:  moduleID,
+		New: func() caddy.Module { return new(SouinCaddyPlugin) },
+	}
+}
+
+type getterContext struct {
+	rw   http.ResponseWriter
+	req  *http.Request
+	next caddyhttp.Handler
+}
+
+// ServeHTTP implements caddyhttp.MiddlewareHandler.
+func (s *SouinCaddyPlugin) ServeHTTP(rw http.ResponseWriter, req *http.Request, next caddyhttp.Handler) error {
+	getterCtx := getterContext{rw, req, next}
+	ctx := context.WithValue(req.Context(), getterContextCtxKey, getterCtx)
+	req = req.WithContext(ctx)
+	buf := s.bufPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	defer s.bufPool.Put(buf)
+	combo := ctx.Value(getterContextCtxKey).(getterContext)
+	plugins.DefaultSouinPluginCallback(rw, req, s.Retriever, s.RequestCoalescing, func(_ http.ResponseWriter, _ *http.Request) error {
+		recorder := httptest.NewRecorder()
```

Comment:
Recorders shouldn't be used outside of the tests.

Reply:
If it helps, here is a recorder that is more optimized for production use: https://pkg.go.dev/github.com/caddyserver/caddy/v2@v2.4.3/modules/caddyhttp#ResponseRecorder